### ISSUE TITLE
distage-testkit-scalatest: to work around `sbt testOnly` launching all the tests, add `-- -z ""` parameter at the end

### ISSUE THREAD

**Issue by neko-kai** [7mind/izumi#1055]
# Upstream fix

Upstream was fixed long ago, but still not published: https://github.com/scalatest/scalatest/issues/2375

# Current workaround

> I've found that you may recover testOnly with the default implementation by adding  -- -z "" boilerplate at the end! e.g. sbt> testOnly abc.MyTest -- -z "" - would behave just as expected, running only MyTest.

# Problem 

`sbt testOnly` doesn't work correctly with distage-testkit-scalatest – ALL tests run, instead of just the selected.

# Workaround

Instead of regular invocation

```
sbt> testOnly org.mypackage.MyTestClass
```

Add options `-- -z ""` at the end of `testOnly` command as follows:

```
sbt> testOnly org.mypackage.MyTestClass -- -z ""
```

Then `testOnly` works as expected

# Details

As detailed in https://gitter.im/7mind/izumi?at=5ebbf553af599706dff55531 `sbt testOnly` doesn't work for distage-testkit-scalatest tests – because Scalatest does not provide data to suites about which _other_ suites it's going to run – and in sbt-mode there's no way to figure that out via side-channels, like distage-testkit does in Intellij IDEA / vanilla ScalatestRunner (single suite launch works there) there's no choice but to run _all_ the suites.

We can allow the user to workaround that by applying the pattern supplied in `-z` or `-t` to test Suites also - instead of `testOnly AbcTest` write `testOnly -- -n AbcTest`. Additionally it may be possible (but is unlikely) to figure out if we're running in `testOnly` and throw an error in that case, advising a workaround.

related issue: we should open an issue to notify upstream and/or have an issue to link to.

**Comment by neko-kai** [comment]
Correction – actually '-z' doesn't work either, it works only in _conjunction_ with `testOnly <className>` because scalatest passes different filters for each suite - names of tests that pass the filter based on the output of `.testNames` method, so if a different test suite is selected randomly as the runner even `-z` wouldn't run the test case

**Comment by coreyoconnor** [comment]
The "workaround" I've implemented for our project is to override aspects of `DistageScalatestTestSuiteRunner`: Only the suite which has `run` called is executed; No part of `DistageTestsRegistrySingleton` is used.

This resulted in the removal of the parallel execution of Suites. However ,we've found that parallel execution of suites added less value to our dev experience than `testOnly`.

The particular implementation is fragile. A change to the runner could easily break this. However, with some changes to the runner I can imagine providing two mixins: One that uses the parallel suite execution as currently supported; another executing suites traditionally to support `testOnly`.

**Comment by neko-kai** [comment]
@coreyoconnor 
I've found that you may recover `testOnly` with the default implementation by adding ` -- -z ""` boilerplate at the end! e.g. `sbt> testOnly abc.MyTest -- -z ""` - would behave just as expected, running only MyTest.

I think this should be fixed on the scalatest side – there are two options that would fix it:
  1. create test classes first before executing them, which allows discovery via side-effect – same as other scalatest runners, except sbt's, that's why tests under Intellij works correctly
  2. Provide direct information to Suite about other suites – not via side-effect – so we don't have to do discovery using classgraph.
The issue is finding the time to communicate the above or push the implementation to scalatest maintainers (which may be easier now than when the issue was created, since it's a lot more maintained now.

Unfortunately not only parallel execution is lost when running tests separately, but also global memoization – which is the primary reason for having a single point of entry, since scalatest can already run tests in parallel by itself.
Instead of a mixin, a more robust option maybe to simply provide an alternative Runner for scalatest tests that fixes the issues above, so it may be used when necessary as follows, until there are fixes in upstream:
```scala
ThisBuild / testFrameworks -= TestFrameworks.ScalaTest
ThisBuild / testFrameworks += TestFramework("izumi.distage.AltScalatestFramework")
```

**Comment by coreyoconnor** [comment]
I definitely like the idea of having a replacement test framework. That would enable validating changes that could be merged into scalatest without actually having to change scalatest. I'm not opposed to changing scalatest but requiring an merged and updated scalatest to validate these changes would be costly.

Providing direct information to a Suite about the entire test execution request  (EG: what other suites are included) I bet would have the biggest impact. Tho definitely seems the most difficult.
